src/fsspec_xrootd/xrootd.py

```diff
@@ -704,16 +716,16 @@ def __init__(
 
         self.fs = fs
 
-        if "r" in mode:
+        if "r" in mode and self.fs.locate_all_sources:
             self._hosts = self._locate_sources(path)
         else:
             self._hosts = [fs.storage_options["hostid"]]
 
         # Try hosts until you find an openable file
-        for i_host in range(len(self._hosts)):
+        for _i_host in range(len(self._hosts)):
             self._myFile = client.File()
             status, _n = self._myFile.open(
-                fs.protocol + "://" + self._hosts[i_host] + "/" + path,
+                fs.unstrip_protocol(path),
```

Comment:
Ok I think my previous comment was errant: you wanted to cycle through specific hosts rather than use the redirector that was specified when the filesystem was initially constructed. So you had it correct before. Sorry about that.

That said, I think we want to try letting the redirector send us to a host right away and set the initial host we discovered
```
self._hosts = [self._myFile.get_property("DataServer")]
```
Then later we would speculatively re-open the file through the redirector url (i.e. the value of `self.fs.hostid`) with the magic `?tried={','.join(self._hosts)}` appended to the URL to force the redirector to give us a new data server. This is the magic that is in https://github.com/cms-sw/cmssw/blob/master/Utilities/XrdAdaptor/src/XrdRequestManager.cc#L645-L651
We could leave that optionally to a future PR and revert this to the previous implementation.

Reply:
For now, I'll revert this to the previous implementation. Just so I understand, it seems like your suggestion here with the CMSSW magic is only different from the `locate()` implementation in that the redirector may suggest sites in a more optimal way when we request to `open`, as opposed to when we ask it to `locate`. I guess your suggestion also only has us look for new sites when a previous site fails, so that could save us site-searching time.

Thanks for pointing out the URL-appended syntax! It's definitely useful to know that exists.